### PR TITLE
initial svds support based on eigs

### PR DESCRIPTION
**Pull request by jaak-s**
Added support for sparse truncated SVD based on `eigs`. Keyword inputs are the same as for `eigs`. Returns left singular vectors, singular values, right singular vectors (and also outputs from eigs). Added a test case.

### CODE REVIEW

**Comment by ViralBShah** [comment]
Bump @jaak-s

**Comment by jaak-s** [comment]
Thanks for the bump, got busy during the holidays. I'll try to put patches together in the next couple of days.

I tried out the `[0 A; A' 0]` approach you and @Jutho outlined and it seems to work well.

**Comment by jaak-s** [comment]
Now `svds` uses `[0 A; A' 0]` approach, saves type information in `SVDOperator` and works also with complex operators. The doc has not been updated. Plan is to finalize the docs after you all are fine with the patch.

Main changes:
- using approach by @ViralBShah to extract singular vectors
- changed input parameter `nev` to `nsv` (number of singular values)
- `svds` only returns largest singular values. The option for the smallest singular values is currently removed because `[0 A; A' 0]` creates extra small singular values and post-processing is not so clear. If someone knows the exact steps let me know, if not we can leave it for future ToDo.
- When using duck typed `X` in `svds(X)` following methods are required: `size(X)`, `eltype(X)`, `X * vector`, `X' * vector`.

File: base/linalg/arnoldi.jl [hunk 1]
```diff
@@ -106,3 +106,51 @@ function eigs(A, B;
...
+    return Complex128
+  end
+  if et <: Float32
+    return Float32
+  end
+  error("Element type $et is not supported for 'svds'.")
+end
+
+function svds{S}(X::S; nsv::Int = 6, ritzvec::Bool = true, tol::Float64 = 0.0, maxiter::Int = 1000)
+  otype = getOperatorType(X)
```
Comment (ViralBShah) [inline base/linalg/arnoldi.jl]: I think we can just simplify this to use just `eltype(X)` and avoid `getOperatorType()` altogether. The checks for whether eltype() is BlasFloat should ideally be in `eigs`. It should be possible to make `SVDOperator` free of types then.

**Comment by ViralBShah** [comment]
For good performance, we should be using `A_mul_B!`, but that requires some work on `eigs`. For now, nothing much can be done here.

File: base/linalg/arnoldi.jl [hunk 2]
```diff
@@ -106,3 +106,51 @@ function eigs(A, B;
...
+  ex    = eigs(SVDOperator{otype,S}(X), I; ritzvec = ritzvec, nev = 2*nsv, tol = tol, maxiter = maxiter)
+  ind   = [1:2:nsv*2]
+  sval  = abs(ex[1][ind])
+
```
Comment (tkelman) [inline base/linalg/arnoldi.jl]: travis doesn't like the trailing whitespace here or on line 151

**Comment by ViralBShah** [comment]
We also need to filter out the singular values that are zero or close to zero.

**Comment by jaak-s** [comment]
The reason for introducing `getOperatorType` was to support (sparse) binary and integer matrices. Their eltype is bool/integer but they can be multiplied by float vectors. Basically changing the operator type to Float64 for integer matrix allows `eigs` to work on it. Note that `svd` works with integer matrices and I found it is useful to have out-of-the-box support for them in `svds` too (also because in my work I use `svds` on binary matrices).

If you wish I can remove `getOperatorType`.

White-space should be now fixed.

**Comment by jaak-s** [comment]
Regarding removing close to zero singular values, I would prefer the current behavior. I think users like to get N values when they set `nsv=N`, even if last of them are 0. This is also the behavior of `svds` in Octave and in Scipy (`scipy.sparse.linalg.svds`).

**Comment by ViralBShah** [comment]
I agree the integer matrix support is good to have. I suggest that instead of `getOperatorType` in `svds`, we should fix this for `eigs`. To support integer matrices, it is better to convert them to float in `eigs`, since the matrix-vector products are significantly faster then. Would be great if you can add it to `eigs` here, or I can do it later.

**Comment by jaak-s** [comment]
Sounds good, I will replace `getOperatorType` with `eltype` then. Since I am not very familiar with code of `eigs` I'd prefer to leave it to you.

**Comment by ViralBShah** [comment]
About the filtering out of singular values that are 0 - octave does do it. But, we can do it once we get this basic piece of code checked in - once you remove getOperatorType.

**Comment by jaak-s** [comment]
`getOperatorType` is now removed, also added check for `nsv`.

I wonder what singular values does Octave remove because they return 0-valued singular values (version 3.8.1):
```octave
octave:1> svds( zeros(4,3), 2 )
ans =

   0
   0
```

**Comment by ViralBShah** [comment]
This was in the octave source that I quoted above. Of course there are legitimate zero singular values. Reading this carefully, they are doing a filter on negative values, and perhaps something in the case where sigma is specified. I will look deeper. 

````
    ## We wish to exclude all eigenvalues that are less than zero as these
    ## are artifacts of the way the matrix passed to eigs is formed. There
    ## is also the possibility that the value of sigma chosen is exactly
    ## a singular value, and in that case we're dead!! So have to rely on
    ## the warning from eigs. We exclude the singular values which are
    ## less than or equal to zero to within some tolerance scaled by the
    ## norm since if we don't we might end up with too many singular
    ## values.
````

**Comment by ViralBShah** [comment]
Travis failure on linux appears unrelated.

**Comment by ViralBShah** [comment]
I think this is a great start with `svds` and we will refine it as we go along.

**Comment by ViralBShah** [comment]
Added to NEWS in 47061a55ba0ed7e0ed892e54a89faaa94486a923

**Comment by jaak-s** [comment]
Thank you.

I forgot to update the docs earlier but now pushed a commit that reflects the latest version of `svds`. Can you merge it again or should I create a new pull request?

**Comment by ViralBShah** [comment]
It looks like your last commit here is still 8 days ago. Could you create a new PR?

Also, we should have squashed these commits, but it escaped my mind before merging.

**Comment by jaak-s** [comment]
Created PR https://github.com/JuliaLang/julia/pull/9723 for docs.

**Pull request merged**